### PR TITLE
[GSDESCARTE-7] Handle errors in layer loading process

### PR DESCRIPTION
**Pull request by tkohr**
PR ascends (tile/image loading errors) into job status (by layer and tile) and allows to activate a `debug` mode per layer to render erroneous tiles pink.

Can be tested with erroneous layers, e.g. copying test data into demo directory and using larger extent:

```
{
  "layers": [
    {
      "type": "XYZ",
      "url": "/data/osm-tiles/{z}/{x}/{y}.png"
    }
  ],
  "size": [1024, 1024],
  "center": [12, 48],
  "dpi": 50,
  "scale": 40000000,
  "projection": "EPSG:3857"
}
```

TODOs:

- [x] show missing tiles with a specific color in the final image => can be configured via `debug` param in layer spec
- [x] still give an image to the user even if one or more layers fail => has already been the case
- [x] provide the user with the necessary information regarding errors => should raise an error state for a tile / layer in the job
- [x] deblock non-tiled WMS for `debug: false`
- [ ] add and adapt tests (observable completion)

### CODE REVIEW

File: src/printer/job.js [hunk 1]
```diff
@@ -45,33 +45,42 @@ export function createJob(spec) {
     .pipe(
       switchMap((layerStates) => {
         const allReady = layerStates.every(([progress]) => progress === 1);
-
+        let errors = [];
         if (allReady) {
           for (let i = 0; i < layerStates.length; i++) {
             context.drawImage(layerStates[i][1], 0, 0);
+            if (layerStates[i][2] && layerStates[i][2].length > 0) {
```
Comment (fgravin) [inline src/printer/job.js:52]: use explicit variable
```js
const canvasImage = layerStates[i][1]
const errors = layerStates[i][2]
```

File: src/printer/job.js [hunk 2]
```diff
@@ -45,33 +45,42 @@ export function createJob(spec) {
     .pipe(
       switchMap((layerStates) => {
         const allReady = layerStates.every(([progress]) => progress === 1);
-
+        let errors = [];
```
Comment (fgravin) [inline src/printer/job.js:48]: Should be typed

File: src/printer/job.js [hunk 3]
```diff
@@ -45,33 +45,42 @@ export function createJob(spec) {
...
         }
       }),
-      map(([progress, imageBlob]) => {
+      map(([progress, imageBlob, errors]) => {
         return {
           ...job,
           progress,
           imageBlob,
           status: progress === 1 ? 'finished' : 'ongoing',
+          errors: errors,
```
Comment (fgravin) [inline src/printer/job.js:83]:
```suggestion
          errors,
```

File: src/printer/layers.js [hunk 4]
```diff
@@ -39,14 +43,15 @@ export function createLayer(layerSpec, rootFrameState) {
...
-function createTiledLayer(source, rootFrameState, opacity) {
+function createTiledLayer(source, rootFrameState, opacity, debug) {
   const width = rootFrameState.size[0];
   const height = rootFrameState.size[1];
   const context = createCanvasContext2D(width, height);
   context.canvas.style = {};
   let frameState;
   let layer;
   let renderer;
+  let errors = [];
```
Comment (fgravin) [inline src/printer/layers.js:54]: type

File: src/printer/layers.js [hunk 5]
```diff
@@ -1,14 +1,18 @@
...
 import { createCanvasContext2D } from 'ol/dom';
-import { BehaviorSubject, interval } from 'rxjs';
+import { BehaviorSubject, interval, combineLatest } from 'rxjs';
 import { map, startWith, takeWhile, throttleTime } from 'rxjs/operators';
 import { isWorker } from '../worker/utils';
+import { HTTP_CODES } from '../shared/constants';
 
 const update$ = interval(500);
+let errorTile$ = new BehaviorSubject();
+let errorWMS$ = new BehaviorSubject();
```
Comment (fgravin) [inline src/printer/layers.js:15]: no need for observables here

File: src/printer/layers.js [hunk 6]
```diff
@@ -55,18 +60,35 @@ function createTiledLayer(source, rootFrameState, opacity) {
...
       image.hintImageSize(tileSize[0], tileSize[1]);
     }
 
-    image.src = src;
+    var xhr = new XMLHttpRequest();
+    xhr.responseType = 'blob';
+    xhr.addEventListener('loadend', function () {
+      var data = this.response;
+
+      if (this.status === 0 || this.status >= 400) {
```
Comment (fgravin) [inline src/printer/layers.js:79]: set TileState.error here

File: src/printer/layers.js [hunk 7]
```diff
@@ -55,18 +60,35 @@ function createTiledLayer(source, rootFrameState, opacity) {
...
+        if (debug) {
+          createErrorTile(image, tileSize, this.status);
+        }
+        errorTile$.next({ origin: tile.key, tile: tile, error: this.status });
```
Comment (fgravin) [inline src/printer/layers.js:83]: not sure we need the size or the tile ..

Comment (fgravin) [inline src/printer/layers.js:83]: origin should be the tile url, not the key

File: src/printer/layers.js [hunk 8]
```diff
@@ -100,14 +122,30 @@ function createTiledLayer(source, rootFrameState, opacity) {
...
+        error.tile.setState(TileState.ERROR);
+      }
+      // check if error is not already present from previous interval
+      if (errors[errors.length - 1] !== error) {
+        // check if error comes from same source
+        if (source.key_ === error.tile.key) {
+          errors.push(error);
+        }
+      }
+
```
Comment (fgravin) [inline src/printer/layers.js:148]: remove

File: src/printer/layers.js [hunk 9]
```diff
@@ -123,15 +161,50 @@ function createTiledLayer(source, rootFrameState, opacity) {
...
       } else {
-        return [progress, null];
+        return [progress, null, errors];
       }
     }),
     throttleTime(500, undefined, { leading: true, trailing: true })
   );
 }
 
+function createErrorTile(image, tileSize, errorCode) {
```
Comment (fgravin) [inline src/printer/layers.js:173]: move to printer/utils

Comment (fgravin) [reply src/printer/layers.js:173]: type definition

**Review by fgravin — CHANGES_REQUESTED** [review]
Good approach but i think it's too complex for nothing, errors don't have to be an observable.

**Comment by tkohr** [comment]
@fgravin thanks a lot for your feedback. I have opened a much simpler PR regarding this issue here https://github.com/camptocamp/inkmap/pull/16

**Comment by fgravin** [comment]
continued in https://github.com/camptocamp/inkmap/pull/16

**Pull request closed without merging**